src/python/pants/backend/docker/rules.py

```diff
@@ -0,0 +1,330 @@
+    logger.debug(f"Packages for Docker image: {packages_str}")
+
+    digest = await Get(Digest, MergeDigests(artifacts))
+
+    return DockerPackages(artifacts=digest, built=built_packages)
+
+
+# -----------------------------------------------------------------------------------------------
+# Docker binary.
+# -----------------------------------------------------------------------------------------------
+
+
+class DockerBinary(BinaryPath):
+    """The `docker` binary."""
+
+    DEFAULT_SEARCH_PATH = SearchPath(("/usr/bin", "/bin", "/usr/local/bin"))
+
+    def build_image(
+        self, tag: str, digest: Digest, context_root: str, dockerfile: Optional[str] = None
+    ):
+        args = [self.path, "build", "-t", tag]
+        if dockerfile:
+            args.extend(["-f", dockerfile])
+        args.append(context_root)
+
+        return Process(
+            argv=tuple(args),
+            input_digest=digest,
+            description=f"Building docker image {tag}",
+        )
+
+
+@dataclass(frozen=True)
+class DockerBinaryRequest:
+    rationale: str
+    search_path: SearchPath = DockerBinary.DEFAULT_SEARCH_PATH
+
+
+@rule(desc="Finding the `docker` binary", level=LogLevel.DEBUG)
+async def find_docker(docker_request: DockerBinaryRequest) -> DockerBinary:
+    request = BinaryPathRequest(
+        binary_name="docker",
+        search_path=docker_request.search_path,
+        test=BinaryPathTest(args=["-v"]),
+    )
+    paths = await Get(BinaryPaths, BinaryPathRequest, request)
+    first_path = paths.first_path
+    if not first_path:
+        raise BinaryNotFoundError(request, rationale=docker_request.rationale)
+    return DockerBinary(first_path.path, first_path.fingerprint)
+
+
+# -----------------------------------------------------------------------------------------------
+# Build Docker image.
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class DockerBuildContext:
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class DockerBuildContextRequest:
+    address: Address
+    context_root: str
+    targets: Tuple[Target, ...]
+
+
+@rule
+async def create_docker_build_context(request: DockerBuildContextRequest) -> DockerBuildContext:
+    # Get all sources.
+    sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            sources_fields=[t.get(Sources) for t in request.targets],
+            for_sources_types=(DockerImageSources, FilesSources, ResourcesSources),
+        ),
+    )
+
+    # Get all dependent targets.
+    dependencies_targets = await MultiGet(
+        Get(Targets, DependenciesRequest(t.get(Dependencies))) for t in request.targets
+    )
+
+    # Get all packages from those targets.
+    dependencies_packages = await MultiGet(
+        Get(DockerPackages, Targets, dep_targets) for dep_targets in dependencies_targets
+    )
```

Comment:
I think you can do something like:

```
embedded_pkg_tgts = ...  # Figure out which targets are relevant
field_sets_per_tgt = await Get(
    FieldSetsPerTarget, FieldSetsPerTargetRequest(PackageFieldSet, embedded_pkg_tgts)
)
embedded_packages = await MultiGet(
    Get(BuiltPackage, PackageFieldSet, field_set) for field_set in field_sets_per_tgt.field_sets
)
```

When figuring out how to populate `embedded_pkg_tgts`, there may be some minor complication over the fact that `DockerFieldSet` is in the`PackageFieldSet` union itself. So you'll have to exclude any dependencies that are `DockerFieldSet` from the list of embedded_pkg_tgts (we don't want to embed docker images inside other docker images, right? presumably a dep on a docker image means we extend that image in our Dockerfile?). This will have the good side-effect of excluding the original target itself.

One way to do this is to have a `BinaryFieldSet` as a subclass of `PackageFieldSet`, and here we only embed binaries? There are other package types that also don't make sense to embed in a docker image, such as lambdas, and python dists.  

I will let @Eric-Arellano weigh in, as I may be wrong about all this.

Reply:
The input to this rule is a list of targets to embed already.. so it is in the build docker image rule that determines which targets are relevant. And it uses all transitive targets.. so, from what I can tell, it would be all targets (and sources for those) but not the current target itself.

> 13:41:25.59 [DEBUG] Get packages for Docker image from: testprojects/src/python/hello/main, testprojects/src/python/hello/main/main.py:lib, testprojects/src/python/hello/greet/greet.py, 3rdparty/python:ansicolors, 3rdparty/python:requirements.txt

And, when requesting the `PackageFieldSet` for those targets, only the relevant targets are used, from what I can tell, so I get this:

> 13:41:25.61 [DEBUG] Packages for Docker image: testprojects.src.python.hello.main/main.pex

So it feels like the filtering for applicable targets work in the rule producing the `FieldSetsPerTarget`, based on the requested FieldSet.